81-question-mark-chrysalis/technical_specification.md

```diff
@@ -0,0 +1,230 @@
+
+
+#### User Journeys: Service Consumer Receives An Original AnnotatedEMPack
+
+The transformation operation is triggered when the service consumer receives "original annotated Em Pack" or a “re-transform metadata” event.
+
+1. Fetch the datapack ids and corresponding schema ids of all datapacks in the collection where the original_id field matches our original datapack id. Create a mapping of these schema ids to their datapack ids that we call “dirty map”, because all of these datapacks need to be either re-created or deleted.
+2. Create a “transformed map” that maps schema ids to datapacks. This will hold the datapacks that have been transformed in this operation already. Initialize it with just the original schemapack id mapping to the original datapack.
+3. Traverse the schemas in the transformation graph starting with the original schema, in topological order. For all of these schemas, do the following:
+   1. Get a route that has the current schema as input. Since we do not allow multiple routes between the same two schemas, there should be at most one such route. Raise an error if there are multiple such routes.
+   2. Get the datapack from the “transformed map” that corresponds to the input schema of that route. This is the “input datapack” for this step. It should always exist at this point if the topological order was computed correctly, and we can raise an error at this point if this is not the case.
+   3. Compute the transformed datapack using the input datapack and the workflow route.
+   4. If the current schema id is in the “dirty map”, remove it there and update the “computed map” with the transformed datapack.
+   4. Otherwise, add the transformed datapack to the “computed map” with a newly generated id, setting its schema_id to the current schema id, and its original_id field to the original datapack id.
+4. Finally, upsert all entries in the “transformed map” to the database that should be published, and delete all remaining entries in the “dirty map” from the database.
+
+This course of action avoids deleting “dirty” datapacks while they are recreated, since that would make the corresponding resources disappear while the transformation is ongoing. It also keeps any intermediate datapacks in memory, avoiding unnecessary database operations and operations.
+
+#### User Journeys: Configuration Change
+
+The “configuration change” operation should be triggered whenever anything in the Schemas, Workflows or WorkflowRoutes collection changes. It will be good to collect such changes and make them all at once, before triggering this operation, because it is very costly.
+
+The configuration change will trigger the following operations:
+1. Validate the transformation configuration
+2. Recalculate the topological order of the schemas and update the `order` field
+3. Re-compute the schemapacks of all transformed schemas
+4. Re-run the transformations for all original datapacks
+
```

Comment:
I think the first journey (Schema Derivation) and this one could be a single one that triggers whenever there is a transformation config change.